### PR TITLE
Fix #11161: Apply cargo filter to shared groups.

### LINKED ISSUE #11161: [Bug]: Vehicle list cargo filter ignored when grouping by shared orders
### Version of OpenTTD

13.4

### Expected result

Using the cargo filter dropdown in a vehicle group only shows vehicles which can load that cargo.

### Actual result

oil

All vehicles in the selected group are shown, regardless of what they can carry.

This is true for all types of vehicles, not just RVs.

### Steps to reproduce

1. Buy some vehicles fitted for different cargos (for example, Oil and Passengers)
2. Filter the list to show only vehicles that carry Oil
3. Observe that the Passengers vehicles disappear from the list
4. Group the vehicle list by shared orders
5. Observe that the Passengers vehicles reappear

### PR DESCRIPTION
**Pull request by Durchbruchswagen**
## Motivation / Problem
Fix #11161 

Closes #11294

## Description
To solve the problem I had to somehow apply a cargo filter on groups of vehicles, unfortunately currently used filter only works for groups of size one (so basically single vehicle). I workaround it by first sorting all vehicles by shared orders and then gathering them into vehicle groups of size one. After that I applied a cargo filter and was left with vector of filtered vehicle groups of size one, sorted by their shared orders so I iterated over this vector gathering those single vehicles into groups.

## Limitations
During testing I noticed that even though shared groups don't show vehicles that don't carry a specified cargo type when you inspect one of those groups you see all vehicles that share that certain order, even those that don't carry specified cargo type.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/vehicle_gui.cpp [hunk 1]
```diff
@@ -203,23 +203,34 @@ void BaseVehicleListWindow::BuildVehicleList()
 		this->FilterVehicleList();
 	} else {
 		/* Sort by the primary vehicle; we just want all vehicles that share the same orders to form a contiguous range. */
-		std::stable_sort(this->vehicles.begin(), this->vehicles.end(), [](const Vehicle * const &u, const Vehicle * const &v) {
+		std::stable_sort(this->vehicles.begin(), this->vehicles.end(), [](const Vehicle * const &u, const Vehicle * const &v){
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp]:
```suggestion
		std::stable_sort(this->vehicles.begin(), this->vehicles.end(), [](const Vehicle * const &u, const Vehicle * const &v) {
```

File: src/vehicle_gui.cpp [hunk 2]
```diff
@@ -203,23 +203,34 @@ void BaseVehicleListWindow::BuildVehicleList()
...
+			if(u->FirstShared() == v->FirstShared()){
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp]:
```suggestion
			if (u->FirstShared() == v->FirstShared()) {
```

File: src/vehicle_gui.cpp [hunk 3]
```diff
@@ -203,23 +203,34 @@ void BaseVehicleListWindow::BuildVehicleList()
...
+
+		uint max_num_vehicles = 0;
 
-			this->vehgroups.emplace_back(begin, end);
+		GUIVehicleGroupList::iterator begin = this->vehgroups.begin();
+		GUIVehicleGroupList::iterator end = begin;
+		std::advance(end, 1);
 
+		while (begin != this->vehgroups.end()) {
+			while(end != this->vehgroups.end() && begin->vehicles_begin[0]->FirstShared()==end->vehicles_begin[0]->FirstShared()){
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp]:
```suggestion
			while (end != this->vehgroups.end() && begin->vehicles_begin[0]->FirstShared() == end->vehicles_begin[0]->FirstShared()) {
```

File: src/vehicle_gui.cpp [hunk 4]
```diff
@@ -203,23 +203,34 @@ void BaseVehicleListWindow::BuildVehicleList()
...
+				begin->vehicles_end = end->vehicles_end;
+				end=this->vehgroups.erase(end);
```
Comment (2TallTyler) [inline src/vehicle_gui.cpp]:
```suggestion
				end = this->vehgroups.erase(end);
```

**Review by 2TallTyler — COMMENTED** [review]
No comments on the actual fix, but there are a bunch of missing spaces. 🙂

**Comment by TrueBrain** [comment]
What you purpose is a fine solution. What triggered me a bit was the limitation. I would expect that if you filter the list of shared orders, you will see the full group when only one vehicle matches. I would not expect the group to be filtered itself. This depends a bit on your definition of the filtering .. I would expect the filtering to be done after-the-fact, while you suggest to do it before-the-fact. This is just an opinion, and there isn't a right or wrong here.

I am curious, what method do you prefer? As an alternative, I wrote https://github.com/OpenTTD/OpenTTD/pull/11294 . Which does what I suggest. I am very curious to your opinion in the matter.

**Comment by Durchbruchswagen** [comment]
So I checked your solution and I think that I like yours a little bit more than mine, to be honest. Besides being less awkward than mine,  filtering after-the-fact seems more "intuitive" because if someone further inspects the vehicle group they are going to see the entire group and by that I mean they are going to see even those vehicles that do not carry specified cargo but share the same order. 
Small little nitpick to your solution. In `CargoFilter` after you set `match` to `true` there is no way that it is going to turn back to `false` so you may as well `return true` at this point.

**Pull request closed without merging**